packages/node-core/src/indexer/multiChainRewind.service.ts

```diff
@@ -0,0 +1,272 @@
+import {IBlockchainService} from '../blockchain.service';
+import {NodeConfig} from '../configure';
+import {createRewindTrigger, createRewindTriggerFunction, getPgPoolConfig, getTriggers} from '../db';
+import {MultiChainRewindEvent, MultiChainRewindPayload} from '../events';
+import {getLogger} from '../logger';
+import {StoreService} from './store.service';
+import {PlainGlobalModel} from './storeModelProvider/global/global';
+import {Header} from './types';
+
+const logger = getLogger('MultiChainRewindService');
+
+export enum RewindStatus {
+  /** The current chain is in normal state. */
+  Normal = 'normal',
+  /** The current chain is waiting for other chains to rewind. */
+  WaitOtherChain = 'waitOtherChain',
+  /** The current chain is executing rewind. */
+  Rewinding = 'rewinding',
+}
+export interface IMultiChainRewindService {
+  chainId: string;
+  status: RewindStatus;
+  waitRewindHeader?: Header;
+  setGlobalRewindLock(rewindTimestamp: number): Promise<void>;
+  /**
+   * Check if the height is consistent before unlocking.
+   * @param tx
+   * @param rewindTimestamp The timestamp to roll back to, in milliseconds.
+   * @returns the number of remaining rewind chains
+   */
+  releaseChainRewindLock(tx: Transaction, rewindTimestamp: number): Promise<number>;
+}
+
+export interface IMultiChainHandler {
+  handleMultiChainRewindEvent(rewindBlockPayload: MultiChainRewindPayload): void;
+}
+
+/**
+ * Working principle:
+ * multiChainRewindService is primarily responsible for coordinating multi-chain projects.
+ * When global.rewindLock changes, a PG trigger sends a notification, and all subscribed chain projects will receive the rollback notification.
+ * This triggers a rollback process, where the fetch service handles the message by clearing the queue.
+ * During the next fillNextBlockBuffer loop, if it detects the rewinding state, it will execute the rollback.
+ */
+@Injectable()
+export class MultiChainRewindService implements IMultiChainRewindService, OnApplicationShutdown {
+  private _status: RewindStatus = RewindStatus.Normal;
+  private _chainId?: string;
+  private _dbSchema?: string;
+  private _rewindTriggerName?: string;
+  private pgListener?: PoolClient;
+  private _globalModel?: PlainGlobalModel = undefined;
+  waitRewindHeader?: Header;
+  constructor(
+    private nodeConfig: NodeConfig,
+    private eventEmitter: EventEmitter2,
+    private sequelize: Sequelize,
+    private storeService: StoreService,
+    @Inject('IBlockchainService') private readonly blockchainService: IBlockchainService
+  ) {}
+
+  private set chainId(chainId: string) {
+    this._chainId = chainId;
+  }
+
+  get chainId(): string {
+    assert(this._chainId, 'chainId is not set');
+    return this._chainId;
+  }
+
+  private get dbSchema(): string {
+    assert(this._dbSchema, 'dbSchema is not set');
+    return this._dbSchema;
+  }
+
+  private set dbSchema(dbSchema: string) {
+    this._dbSchema = dbSchema;
+  }
```

Comment:
Im not sure why private getters and setters are needed. Can't there just be a private variable?

Reply:
Same for `rewindTriggerName`